### ISSUE TITLE
Spring Cloud AWS - batch listener unable to receive more than 10 messages - fifo queue

### ISSUE THREAD

**Issue by B-hamza** [awspring/spring-cloud-aws#1020]
**Type**: Bug

**Component**: "SQS"

**Describe the bug**
Can't read more than 10 messages in a batch when using multiple groupe Id in fifo queue.
I tested with a queue that contains 10000 message distributed in 40 message groupe Id. 
My listener configurations is :
```
 @SqsListener(value = "product.fifo",
            maxMessagesPerPoll = "40",
            maxConcurrentMessages = "40",
            acknowledgementMode = "MANUAL",
            pollTimeoutSeconds = "30"
    )
    public void listen(List<Message<Product>> messages, BatchAcknowledgement<Product> batchAcknowledgement) {
```
I only reveive 10 messages, tested multiple configuration, maybe i'm doing something wrong.

**Comment by takima-youssoupha** [comment]
Up !

**Comment by takima-youssoupha** [comment]
Up !

**Comment by Sylvain-transparence** [comment]
I got the same problem !

**Comment by tomazfernandes** [comment]
Hi everyone, would any of you be able to create a sample project that reproduces the issue?

Thanks.

**Comment by B-hamza** [comment]
Hello, than you for you response, you can find a sample here https://github.com/B-hamza/sample-sqs 
many thanks

**Comment by tomazfernandes** [comment]
Thanks for the app @B-hamza.

There are two main causes for not getting more than 10 messages in this app.

1. The first reason is the way `FIFO` works. `SQS` only allows a single poll to fetch 10 messages in batch. For a `FIFO` queue, this means that at most 10 messages from a given `message group` will be retrieved, and `FIFO` won't release any further messages from such message group before the previous batch has been either acknowledged or past visibility timeout.

As stated [in the docs](https://docs.awspring.io/spring-cloud-aws/docs/3.1.0/reference/html/index.html#fifo-support) by default the framework will break batches by message group, so you can process messages from **different message groups in parallel**, while **keeping order within each batch**. 

It's possible to override this behavior to allow a batch with different message groups if it makes sense, but we don't have a `ContainerOption` at this time for this, so it'd require some manual wiring to get there, or a PR with adding such option to be used when building the components.

Summing up - for e.g. 40 message groups, your app can receive 400 messages at most at a given time, but they will be processed in parallel in batches of 10 messages per message group.

2. The second reason is - if you send and receive messages at the same time, you won't have enough messages in the queue at a given time to fulfill your 20 or 40 messages quota per batch. The proper way to test this feature is to have let's say 1000 messages in the queue first with the app stopped, then starting the app.

Does this make sense to you all? Any questions or anything I might be missing?

Thanks.

**Comment by B-hamza** [comment]
Thank you Tomaz for your explanations ! 
I already tried for more than 1000 messages in the queue before starting my listener i still get 10 messages only, this was my first test actually as explained in my first post.

**Comment by tomazfernandes** [comment]
@B-hamza, yes, for FIFO queues you won't get more than 10 messages in a batch because that's the maximum SQS allows for a given message group.

What we can do is create a feature request so that it's possible to receive messages from all message groups in the same List.

Would that solve your issue?

Also, if that's the case, can all of you please share what's your use case for needing to receive messages from many message groups in the same batch? Just so I have an idea.

Thanks.

**Comment by tomazfernandes** [comment]
Folks, I'll close this issue in favor of https://github.com/awspring/spring-cloud-aws/issues/1041

If anyone would like to contribute a PR for this, I'll be happy to review.

Thanks.

**Comment by tomazfernandes** [comment]
Closed in favor of https://github.com/awspring/spring-cloud-aws/issues/1041
